### ISSUE TITLE
Performing nmap -p 28015 <ip address> Crashes the rethinkdb process

### ISSUE THREAD

**Issue by marchaos** [rethinkdb/rethinkdb#6084]
This is a DoS vulnerability. Tested against 2.3.5 os x version of rethinkdb. I performed this on a different host the that of the rethinkdb process.

**Comment by danielmewes** [comment]
Thanks for reporting this @marchaos . Does RethinkDB print a backtrace or any other message when it crashes?

How did you install RethinkDB? Through homebrew, or from the .dmg file?

(I can't reproduce this on Linux so far. Will test on OS X later.)

**Comment by danielmewes** [comment]
Ok was able to reproduce this on OS X:
```
Version: rethinkdb 2.3.5 (CLANG 7.3.0 (clang-703.0.31))
error: Error in src/arch/io/network.cc at line 174:
error: Guarantee failed: [res != -1] Could not set SO_KEEPALIVE option.
error: Backtrace:
error: Tue Aug 30 11:17:01 2016

       1 [0x10ad65494]: 0   rethinkdb                           0x000000010ad65494 _Z19rethinkdb_backtracePPvi + 132 at 0x10ad65494 ()
       2 [0x10a8217ac]: 0   rethinkdb                           0x000000010a8217ac _ZN11backtrace_tC2Ev + 76 at 0x10a8217ac ()
       3 [0x10a822ac7]: 0   rethinkdb                           0x000000010a822ac7 _ZN26lazy_backtrace_formatter_tC2Ev + 39 at 0x10a822ac7 ()
       4 [0x10a8212f0]: 0   rethinkdb                           0x000000010a8212f0 _Z16format_backtraceb + 32 at 0x10a8212f0 ()
       5 [0x10aaf4d96]: 0   rethinkdb                           0x000000010aaf4d96 _Z18report_fatal_errorPKciS0_z + 406 at 0x10aaf4d96 ()
       6 [0x10a80d85a]: 0   rethinkdb                           0x000000010a80d85a _ZN16linux_tcp_conn_t16enable_keepaliveEv + 74 at 0x10a80d85a ()
       7 [0x10a873f72]: 0   rethinkdb                           0x000000010a873f72 _ZN14query_server_t11handle_connERK12scoped_ptr_tI27linux_tcp_conn_descriptor_tEN14auto_drainer_t6lock_tE + 178 at 0x10a873f72 ()
       8 [0x10a879468]: 0   rethinkdb                           0x000000010a879468 _ZNSt3__110__function6__funcINS_6__bindIM14query_server_tFvRK12scoped_ptr_tI27linux_tcp_conn_descriptor_tEN14auto_drainer_t6lock_tEEJPS3_RNS_12placeholders4__phILi1EEESA_EEENS_9allocatorISI_EEFvRS6_EEclESL_ + 72 at 0x10a879468 ()
       9 [0x10a812486]: 0   rethinkdb                           0x000000010a812486 _ZN32linux_nonthrowing_tcp_listener_t6handleEi + 54 at 0x10a812486 ()
       10 [0x10a818225]: 0   rethinkdb                           0x000000010a818225 _ZN6coro_t3runEv + 69 at 0x10a818225 ()
error: Exiting.
Trace/BPT trap: 5
```

**Comment by danielmewes** [comment]
We are getting an EINVAL return code from `setsockopt`, which on OS X can mean:
```
[EINVAL]           The socket has been shut down.
```

(On Linux this seems to behave differently, and the man page doesn't list this case.)

We need to handle this result.

**Comment by danielmewes** [comment]
A fix for this is now in branch daniel_6084, and internal code review 3741.

**Comment by marchaos** [comment]
Hi @danielmewes. Just saw this. Looks like it's in hand, so thanks 👍

**Comment by danielmewes** [comment]
Merged into `next` via 4944779 and backported to `v2.3.x` via 3558c75.

Unfortunately this just missed the cut for release 2.3.5. If we do another bug fix release before RethinkDB 2.4.0, this will go into it. Otherwise it will ship with RethinkDB 2.4.

(I don't think this is a super serious DoS issue, because the ReQL client port should already not be exposed to the public. It also is only an issue on OS X, and most production servers run on Windows or Linux. So we'll wait for the next release to ship this.)

**Comment by marchaos** [comment]
Agreed. Fine with me.

**Comment by AtnNn** [comment]
A fix for this issue has been released in [RethinkDB 2.3.6](https://rethinkdb.com/blog/2.3.6-release/)
